### PR TITLE
Stopping criteria: Tolerance in objective function over iterations

### PR DESCRIPTION
**Pull request by nishnash54**
Updated stopping criteria extending the `ftol` functionality.
Stop the optimization if the objective function value does not improve over `ftol_iter` iterations.

## Description

Enhancements:
 - Added a `ftol_iter` parameter to BinaryPSO, GeneralPSO, LocalBestPSO and GlobalBestPSO.
 - The tolerance `ftol` of the objective function must be consistent over the next `ftol_iter` iterations
 - Initial functionality of `ftol` is maintained as the default value of `ftol_iter` is set to `1`

## Related Issue
#373 

## Motivation and Context

The are 3 different stopping criteria defined ( maxiter , minstep, minfunc), all of these are hard set values. The proposed enhancement is often used in metaheuristic optimization, where the `maxiter` cannot be pre determined. In these cases, the user can define the stopping criteria as 

> If the objective function value does not improve for N iterations stop the optimization and return the best objective function value and other related objects.

## How Has This Been Tested?

Tests:
 - New test added to ` tests/optimizers/test_general_optimizer.py`
 - The minimum number of iterations before which the optimizer stops must be at least equal to `ftol_iter`
 - Passes tests related to initial functionality of `ftol`.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ljvmiranda921** [comment]
Hi @nishnash54 ! Thank you so much for this PR! **Let me look at this on Thursday/Friday.** As this seems to be an important feature! :bowing_man: 
Have a good day!

File: pyswarms/base/base_discrete.py [hunk 1]
```diff
@@ -93,6 +101,7 @@ def __init__(
         self.options = options
         self.init_pos = init_pos
         self.ftol = ftol
+        self.ftol_iter=ftol_iter
```
Comment (ljvmiranda921) [inline pyswarms/base/base_discrete.py:104]: Nit: let's add a space between the equal signs for consistency

```
self.ftol_iter = ftol_iter
```

File: pyswarms/discrete/binary.py [hunk 2]
```diff
@@ -197,11 +200,13 @@ def optimize(self, objective_func, iters, n_processes=None, **kwargs):
             self._populate_history(hist)
             # Verify stop criteria based on the relative acceptable cost ftol
             relative_measure = self.ftol * (1 + np.abs(best_cost_yet_found))
-            if (
-                np.abs(self.swarm.best_cost - best_cost_yet_found)
-                < relative_measure
-            ):
-                break
+            delta = np.abs(self.swarm.best_cost - best_cost_yet_found) < relative_measure
```
Comment (ljvmiranda921) [inline pyswarms/discrete/binary.py:203]: Thanks for making this part clearer!

File: pyswarms/single/global_best.py [hunk 3]
```diff
@@ -129,6 +134,7 @@ def __init__(
             velocity_clamp=velocity_clamp,
             center=center,
             ftol=ftol,
+            ftol_iter=ftol_iter,
```
Comment (ljvmiranda921) [inline pyswarms/single/global_best.py:137]: Nit: similar to the one before, please add spaces between the operator:

```python
ftol_iter = ftol_iter
```

File: pyswarms/single/local_best.py [hunk 4]
```diff
@@ -154,6 +159,7 @@ def __init__(
             velocity_clamp=velocity_clamp,
             center=center,
             ftol=ftol,
+            ftol_iter=ftol_iter,
```
Comment (ljvmiranda921) [inline pyswarms/single/local_best.py:162]: Same comment

File: pyswarms/single/global_best.py (hunk 3 above)
Comment (nishnash54) [reply pyswarms/single/global_best.py:137]: This seems to be consistent with the above lines.
Do you recommend changing all the lines? Or is this fine?

Same query regarding the next change.

Comment (ljvmiranda921) [reply pyswarms/single/global_best.py:137]: Oof you’re right! Ha! I didnt notice that! 
Ideally there should be some spacing in between (or to be really sure, what the black formatter gives), but for this PR let’s leave it at that. No action needed for this snippet

**Comment by ljvmiranda921** [comment]
Oof you’re right! Ha! I didn’t notice that!
Ideally we should follow what the black formatter gives,
but let’s leave it for now, that should be a different PR.

Just one code change and we’re good to merge, @nishnash54

**Comment by ljvmiranda921** [comment]
Awesome! Thanks for this @nishnash54 ! Hope to see you around again soon, @all-contributors please add @nishnash54 for code

**Pull request merged**